File: backend/src/services/organisation.service.ts
import { prisma } from '../lib/prisma.ts';
import type { CreateOrganisationInput, UpdateOrganisationInput } from '../schema/organisation.schema.ts';

export const OrganisationService = {
  async createOrganisation(input: CreateOrganisationInput, ownerId: string) {
    const org = await prisma.organisation.create({
      data: {
        name: input.name,
        tier: input.tier || 'starter',
        ownerId,
      },
    });

    return org;
  },

  async getOrganisation(organisationId: string) {
    const org = await prisma.organisation.findUnique({
      where: { id: organisationId },
      include: {
        communities: true,
      },
    });

    if (!org) {
      throw new Error('Organisation not found');
    }

    return org;
  },

  async listOrganisations(ownerId: string) {
    const orgs = await prisma.organisation.findMany({
      where: { ownerId },
      include: {
        communities: true,
      },
      orderBy: { createdAt: 'desc' },
    });

    return orgs;
  },

  async updateOrganisation(organisationId: string, ownerId: string, input: UpdateOrganisationInput) {
    const org = await prisma.organisation.findUnique({
      where: { id: organisationId },
    });

    if (!org) {
      throw new Error('Organisation not found');
    }

    if (org.ownerId !== ownerId) {
      throw new Error('Unauthorized: You do not own this organisation');
    }

    const updated = await prisma.organisation.update({
      where: { id: organisationId },
      data: {
        ...(input.name && { name: input.name }),
        ...(input.tier && { tier: input.tier }),
        ...(input.maxCommunities && { maxCommunities: input.maxCommunities }),
      },
    });

    return updated;
  },

  async deleteOrganisation(organisationId: string, ownerId: string) {
    const org = await prisma.organisation.findUnique({
      where: { id: organisationId },
    });

    if (!org) {
      throw new Error('Organisation not found');
    }

    if (org.ownerId !== ownerId) {
      throw new Error('Unauthorized: You do not own this organisation');
    }

    await prisma.organisation.delete({
      where: { id: organisationId },
    });

    return { success: true };
  },
};

File: mobile_app/lib/features/organisations/services/organisation_service.dart
import '../../../shared/models/organisation_model.dart';
import '../../../core/constants/api_constants.dart';
import '../../../core/services/api_client.dart';
import '../../../core/services/error_handler.dart';

class OrganisationService {
  final ApiClient _apiClient = ApiClient();

  /// Create a new organisation
  /// Returns Organisation object
  /// Throws ApiException on error
  Future<Organisation> createOrganisation({
    required String name,
    required String tier,
  }) async {
    try {
      final response = await _apiClient.post(
        createOrganisationEndpoint,
        body: {
          'name': name,
          'tier': tier,
        },
      );

      // Backend response: { success: true, data: { id, name, tier, maxCommunities, ... } }
      return Organisation.fromJson(response['data']);
    } on ApiException {
      rethrow;
    } catch (e) {
      throw ApiException(
        message: 'Failed to create organisation: ${e.toString()}',
        statusCode: 500,
      );
    }
  }

  /// Get a specific organisation by ID
  /// Returns Organisation object
  /// Throws ApiException on error
  Future<Organisation> getOrganisation(String organisationId) async {
    try {
      final endpoint = getOrganisationEndpoint.replaceFirst(
        '{organisationId}',
        organisationId,
      );
      final response = await _apiClient.get(endpoint);

      return Organisation.fromJson(response['data']);
    } on ApiException {
      rethrow;
    } catch (e) {
      throw ApiException(
        message: 'Failed to fetch organisation: ${e.toString()}',
        statusCode: 500,
      );
    }
  }

  /// List all organisations owned by user
  /// Returns list of Organisation objects
  /// Throws ApiException on error
  Future<List<Organisation>> listOrganisations() async {
    try {
      final response = await _apiClient.get(listOrganisationsEndpoint);

      // Backend response: { success: true, data: [{ id, name, ... }, ...] }
      final data = response['data'] as List<dynamic>? ?? [];
      return data
          .map((org) => Organisation.fromJson(org as Map<String, dynamic>))
          .toList();
    } on ApiException {
      rethrow;
    } catch (e) {
      throw ApiException(
        message: 'Failed to fetch organisations: ${e.toString()}',
        statusCode: 500,
      );
    }
  }

  /// Update an organisation
  /// Returns updated Organisation object
  /// Throws ApiException on error
  Future<Organisation> updateOrganisation({
    required String organisationId,
    required String name,
    required String tier,
  }) async {
    try {
      final endpoint = updateOrganisationEndpoint.replaceFirst(
        '{organisationId}',
        organisationId,
      );
      final response = await _apiClient.put(
        endpoint,
        body: {
          'name': name,
          'tier': tier,
        },
      );

      return Organisation.fromJson(response['data']);
    } on ApiException {
      rethrow;
    } catch (e) {
      throw ApiException(
        message: 'Failed to update organisation: ${e.toString()}',
        statusCode: 500,
      );
    }
  }

  /// Delete an organisation
  /// Throws ApiException on error
  Future<void> deleteOrganisation(String organisationId) async {
    try {
      final endpoint = deleteOrganisationEndpoint.replaceFirst(
        '{organisationId}',
        organisationId,
      );
      await _apiClient.delete(endpoint);
    } on ApiException {
      rethrow;
    } catch (e) {
      throw ApiException(
        message: 'Failed to delete organisation: ${e.toString()}',
        statusCode: 500,
      );
    }
  }
}
